### PR TITLE
Blender: Refactor to new publisher

### PR DESCRIPTION
**Pull request by BigRoy**
## Changelog Description

Refactor Blender integration to use the new publisher

## Additional info

Thanks to tremendous effort by @Sharkitty there's a big start, you can actually make valid publishes already!

There are todos however:

- [x] Convert legacy instances
- [x] Fix #TODO statements
- [x] Refactor Load Blend to use new style creator creation for rigs
- [x] Clear up what Create Review "asset_group" pre_create_data should be doing
- [x] ~~Animation Creator should maybe be a HiddenCreator?~~ Animation instances in Blender seem to be much simpler than in Maya so they are easier to create; also loading rigs doesn't automatically create an animation instance yet it seems. So I believe they should remain visible.

And some additional things to test/validate they work as intended:

- [x] Persist instance + context data correctly on saves (toggle states in publisher should work on reset and scene re-open)
- [x] All validators should report `PublishValidationErrors`
- [x] All creators/publishers should still work
- [x] Optional settings should still work

## Testing notes:

1. Launch Blender
2. Use new publisher
3. Break it and report issues here!

Tested Publish types:

- [x] Action - sidenote: There's a known bug from prior to this PR that [loading action does not work](https://github.com/ynput/OpenPype/pull/5910#issuecomment-1810424531).
- [x] Animation
- [x] BlendScene
- [x] Camera
- [x] Layout
- [x] Model
- [x] Pointcache
- [x] Render
- [x] Review
- [x] Rig

### CODE REVIEW

File: openpype/hosts/blender/api/pipeline.py [hunk 1]
```diff
@@ -47,6 +60,101 @@
 log = Logger.get_logger(__name__)
 
 
+class BlenderHost(HostBase, IWorkfileHost, IPublishHost):
```
Comment (iLLiCiTiT) [inline openpype/hosts/blender/api/pipeline.py:63]: Missing `ILoadHost`. And `ls` should be renamed to `get_containers`.

File: openpype/hosts/blender/plugins/publish/extract_blend.py [hunk 2]
```diff
@@ -14,8 +14,14 @@ class ExtractBlend(publish.Extractor):
...
 
     def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+
         # Define extract output file path
 
+        if not self.is_active(instance.data):
+            return
+
```
Comment (iLLiCiTiT) [inline openpype/hosts/blender/plugins/publish/extract_blend.py:24]: duplicated check
```suggestion
```

**Review by LiborBatek — COMMENTED** [review]
Im really thrilled to see this PR in review stage!! so Congrats to everyone involved in it!!

Besides the things which are working well (like model publishing) I have found issue when publishing ```Layout``` family.

First issue being that when ```Export Layout``` being turned ```On``` ( JSON publish ??) it fails during publishing.

b3D_Layout_Extract


When turned off, ```Layout``` got published but fails during ```Load``` when trying to get the asset into some work file...

```
Traceback (most recent call last):

  File "D:\REPO\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
    result = callback(*args, **kwargs)

  File "D:\REPO\OpenPype\openpype\hosts\blender\api\plugin.py", line 448, in _load[publish-report-231114-13-28.json.txt](https://github.com/ynput/OpenPype/files/13348717/publish-report-231114-13-28.json.txt)
  File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_blend.py", line 94, in _process_data
    assert container, "No asset group found"

AssertionError: No asset group found
```

**Comment by LiborBatek** [comment]
and here is full JSON log for that failure during publishing of the Layout with that option being turned ON.
[publish-report-231114-13-28.json.txt](https://github.com/ynput/OpenPype/files/13348734/publish-report-231114-13-28.json.txt)

File: openpype/hosts/blender/plugins/create/create_blendScene.py [hunk 3]
```diff
@@ -2,13 +2,11 @@
 
 import bpy
 
-from openpype.pipeline import get_current_task_name
-from openpype.hosts.blender.api import plugin, lib, ops
-from openpype.hosts.blender.api.pipeline import AVALON_INSTANCES
+from openpype.hosts.blender.api import plugin, lib
 
 
 class CreateBlendScene(plugin.Creator):
```
Comment (simonebarbieri) [inline openpype/hosts/blender/plugins/create/create_blendScene.py:8]:
```suggestion
class CreateBlendScene(plugin.BaseCreator):
    """Generic group of assets."""

    identifier = "io.openpype.creators.blender.blendScene"
    name = "blendScene"
    label = "Blender Scene"
    family = "blendScene"
    icon = "cubes"

    maintain_selection = False

    def _create(self, instance_node):
        selection = lib.get_selection(include_collections=True)
        for data in selection:
            if isinstance(data, bpy.types.Collection):
                instance_node.children.link(data)
            elif isinstance(data, bpy.types.Object):
                instance_node.objects.link(data)


    def create(
        self, subset_name: str, instance_data: dict, pre_create_data: dict
    ):

        instance_node = super().create(subset_name,
                                       instance_data,
                                       pre_create_data)

        if pre_create_data.get("use_selection"):
            mti = ops.MainThreadItem(self._create, instance_node)
            ops.execute_in_main_thread(mti)

        return instance_node
```

This creator still needs to run in the main thread, because to select collections we need to access the outliner, and it isn't enough to change context, because from the Qt window of the creator we cannot access the main Blender window, AFAIK!

I also changed the creator to inherit from the new Blender Creator and added the identifier.

Comment (BigRoy) [reply openpype/hosts/blender/plugins/create/create_blendScene.py:8]: The creator worked fine for me. Could you pull the branch and give it a go?

**Comment by LiborBatek** [comment]
When tested the ```Animation``` publishing came through but when loading the asset into scene it fails on load when using native "Link BlendScene"

```
Traceback (most recent call last):

  File "D:\REPO\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
    result = callback(*args, **kwargs)

  File "D:\REPO\OpenPype\openpype\hosts\blender\api\plugin.py", line 448, in _load
    nodes = self.process_asset(

  File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_animation.py", line 45, in process_asset
    container = data_to.objects[0]

IndexError: list index out of range
```

When using ```Load alembic``` on ```Animation``` asset it produces no error but just the top locator and not content present.
image

**Comment by LiborBatek** [comment]
@simonebarbieri do you know why @BigRoy  missing completely the new  ```blendScene``` family? I guess its because ```develop``` need to be merged into this PR??

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/create/create_blendScene.py:8]: I've tried from the latest commit 😅 if I don't change the inherited class, I cannot even see the family in the creator!

**Comment by LiborBatek** [comment]
When speaking of ```pointcache``` I get error when trying to publish anything. I have tried both to put just geometry into the ```pointcache``` instance or full character hierarchy as on scrn but neither worked.

image

Will enclose a full JSON log:

[publish-report-231114-15-11.json.txt](https://github.com/ynput/OpenPype/files/13349740/publish-report-231114-15-11.json.txt)

**Comment by LiborBatek** [comment]
```Review``` works fine...
image

**Comment by BigRoy** [comment]
> Im really thrilled to see this PR in review stage!! so Congrats to everyone involved in it!!
> 
> Besides the things which are working well (like model publishing) I have found issue when publishing `Layout` family.
> 
> First issue being that when `Export Layout` being turned `On` ( JSON publish ??) it fails during publishing.
> 
> b3D_Layout_Extract
> 
> When turned off, `Layout` got published but fails during `Load` when trying to get the asset into some work file...
> 
> ```
> Traceback (most recent call last):
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
>     result = callback(*args, **kwargs)
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\api\plugin.py", line 448, in _load
>     nodes = self.process_asset(
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_blend.py", line 145, in process_asset
>     container, members = self._process_data(libpath, group_name)
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_blend.py", line 94, in _process_data
>     assert container, "No asset group found"
> 
> AssertionError: No asset group found
> ```

Sorry, my bad - fixed. Had a missing comma somewhere.

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (Sharkitty) [reply openpype/hosts/blender/plugins/create/create_blendScene.py:8]: You're right about the parent class, it's my bad for missing it. But I don't see why you would need to execute in main thread. What version of Blender are you testing on?

**Comment by BigRoy** [comment]
> When tested the `Animation` publishing came through but when loading the asset into scene it fails on load when using native "Link BlendScene"
> 
> ```
> Traceback (most recent call last):
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
>     result = callback(*args, **kwargs)
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\api\plugin.py", line 448, in _load
>     nodes = self.process_asset(
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_animation.py", line 45, in process_asset
>     container = data_to.objects[0]
> 
> IndexError: list index out of range
> ```
> 
> When using `Load alembic` on `Animation` asset it produces no error but just the top locator and not content present. image

Animation instances usually would get created by loading a rig - right? Is that how you did it as well? (Loading rigs would not work yet actually). Do you happen to have an example blender rig file I could test with?

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/create/create_blendScene.py:8]: 3.6! I get this error without changes:
```
Creator: Blender Scene (<property object at 0x000001A82D6BAD40>)

Error: 'NoneType' object has no attribute 'screen'

Traceback (most recent call last):
  File "C:\Dev\Ynput\OpenPype\openpype\pipeline\create\context.py", line 2036, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\plugins\create\create_blendScene.py", line 27, in create
    selection = lib.get_selection(include_collections=True)
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\api\lib.py", line 317, in get_selection
    selection.extend(get_selected_collections())
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\api\lib.py", line 283, in get_selected_collections
    area for area in bpy.context.window.screen.areas
AttributeError: 'NoneType' object has no attribute 'screen'
```

**Comment by Sharkitty** [comment]
> @simonebarbieri do you know why @BigRoy missing completely the new `blendScene` family? I guess its because `develop` need to be merged into this PR??

As @simonebarbieri pointed out, I set the wrong parent class on the blendScene creator.

**Comment by BigRoy** [comment]
> > @simonebarbieri do you know why @BigRoy missing completely the new `blendScene` family? I guess its because `develop` need to be merged into this PR??
> 
> As @simonebarbieri pointed out, I set the wrong parent class on the blendScene creator.

Thanks, fixed with https://github.com/ynput/OpenPype/pull/5910/commits/fa5638d06bb73d4f327d3de120718b72d70c6b9f

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (Sharkitty) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: What about using `bpy.context.screen` instead?

Comment (BigRoy) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: @simonebarbieri could you try again?

Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: @BigRoy oh let me try again!

**Comment by BigRoy** [comment]
> When speaking of `pointcache` I get error when trying to publish anything. I have tried both to put just geometry into the `pointcache` instance or full character hierarchy as on scrn but neither worked.
> 
> image
> 
> Will enclose a full JSON log:
> 
> [publish-report-231114-15-11.json.txt](https://github.com/ynput/OpenPype/files/13349740/publish-report-231114-15-11.json.txt)

I believe this pointcache export might be fixed with https://github.com/ynput/OpenPype/pull/5910/commits/63ff830d12cdfaf8ed4cee4e681f5c5e650cc76a

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: Oh yes, it works :) @BigRoy That's great! I'll try to break it!

Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: Ah @BigRoy the creator still needs an identifier!

Comment (BigRoy) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: Fixed with https://github.com/ynput/OpenPype/pull/5910/commits/f281da5784f79cacda9730c5b211a77e48bfeb2b ? :)

Comment (BigRoy) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: Also I believe we can remove all the `name` attributes from the new style creators as well, right?

**Comment by LiborBatek** [comment]
When tested the ```Action``` family Publish went ok without issues but again it failed on ```Load```

image

```
Traceback (most recent call last):

  File "D:\REPO\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
    result = callback(*args, **kwargs)

  File "D:\REPO\OpenPype\openpype\hosts\blender\api\plugin.py", line 448, in _load
    nodes = self.process_asset(

  File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_action.py", line 77, in process_asset
    collection.children.link(bpy.data.collections[lib_container])

KeyError: 'bpy_prop_collection[key]: key "sh_010_actionMain" not found'
```

but this could be also issue with the old Publisher aka in latest ```develop``` too @simonebarbieri  what u think?

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: It doesn't work if the outliner is not in the main window, but the error that it throws already warns about that. And this was the same even before the new publisher anyway.
I think it is fine to keep it like this for now, because technically we could have even different selections in different outliners...
image
While the selection of objects is actually on the object, and so independent from the outliner. So for now I would keep it like this, and hope that Blender implements a proper way to select and get selected collections.

**Comment by simonebarbieri** [comment]
> When tested the `Action` family Publish went ok without issues but again it failed on `Load`
> 
> ```
> Traceback (most recent call last):
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
>     result = callback(*args, **kwargs)
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\api\plugin.py", line 448, in _load
>     nodes = self.process_asset(
> 
>   File "D:\REPO\OpenPype\openpype\hosts\blender\plugins\load\load_action.py", line 77, in process_asset
>     collection.children.link(bpy.data.collections[lib_container])
> 
> KeyError: 'bpy_prop_collection[key]: key "sh_010_actionMain" not found'
> ```
> 
> but this could be also issue with the old Publisher aka in latest `develop` too @simonebarbieri what u think?

Yes, the Action family is broken and needs to be fixed, but it's unrelated to this PR!

File: openpype/hosts/blender/api/plugin.py [hunk 4]
```diff
@@ -144,20 +147,196 @@ def deselect_all():
...
+        """
+        if not instance_data:
+            instance_data = {}
+
+        instance_data.update(
+            {
+                "id": "pyblish.avalon.instance",
+                "creator_identifier": self.identifier,
+                "label": subset_name,
+                "task": get_current_task_name(),
```
Comment (iLLiCiTiT) [inline openpype/hosts/blender/api/plugin.py:329]: Why is this here?

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (Sharkitty) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: > Also I believe we can remove all the `name` attributes from the new style creators as well, right?

Yes I don't think there's any use for it with the new publisher.

**Comment by LiborBatek** [comment]
I have pulled new commits and the ```Layout``` is working flawlessly now! Altogether with JSON extraction too...

I did test also with ```Manage``` on loaded assets and all working nicely when speaking of ```Layout``` family.

Will continue with those commits pulled.  I can see also the ```BlendScene``` type in Publisher/Create too...

Great progress!!

File: openpype/hosts/blender/api/plugin.py (hunk 4 above)
Comment (BigRoy) [reply openpype/hosts/blender/api/plugin.py:329]: Well spotted - should be removed I suspect. Investigating.

Comment (BigRoy) [reply openpype/hosts/blender/api/plugin.py]: Removed with https://github.com/ynput/OpenPype/pull/5910/commits/0556986a3e39c45be8c923f18a87981c17163c3a

File: openpype/hosts/blender/plugins/create/create_blendScene.py (hunk 3 above)
Comment (BigRoy) [reply openpype/hosts/blender/plugins/create/create_blendScene.py]: Ok, will mark this thread resolved for now - feel free to unresolve if other issues appear.

**Comment by simonebarbieri** [comment]
I've tested the converter for legacy instances and it works perfectly :)

**Comment by LiborBatek** [comment]
@simonebarbieri  So as I have already informed @BigRoy the only not working Instances so far being ```Action``` and ```Animation```

Other than that good!!

**Comment by BigRoy** [comment]
> * ~Animation Creator should maybe be a HiddenCreator?~ Animation instances in Blender seem to be much simpler than in Maya so they are easier to create; also loading rigs doesn't automatically create an animation instance yet it seems. So I believe they should remain visible.

So I've removed this from the todo list due the reason described.

I assume with that it's just a matter of some more tests runs by those who actually know blender, try to break it... otherwise this should be good to go? :)

> the only not working Instances so far being Action and Animation

Animation instances should be fixed with https://github.com/ynput/OpenPype/pull/5910/commits/b04ae30b821c8a26d68eedf62864100e2fd04989

Action instances - no idea what they are and what needs fixing, but [according to](https://github.com/ynput/OpenPype/pull/5910#issuecomment-1810424531) @simonebarbieri those were broken before this PR.

**Comment by alainxi** [comment]
Thanks so much Sharkitty and Roy for all this !

Shouldn't the "Subsets" word be replaced in the new GUI by "Products", to match Ayon's Web App ?
Subsets_vs_products

**Comment by BigRoy** [comment]
UIs will now pop up to the front if they are behind another window or were minimized - I kept searching for e.g. the loader or publisher, this resolves it in Blender. Checked with @simonebarbieri and decided to put it in this PR. 🤞

File: openpype/hosts/blender/plugins/create/create_workfile.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                task_name, task_name, asset_doc, project_name, host_name
+            )
+
+            current_instance.update(
+                {
+                    "asset": asset_name,
+                    "task": task_name,
+                    "subset": subset_name,
+                }
+            )
```
Comment (simonebarbieri) [inline openpype/hosts/blender/plugins/create/create_workfile.py:83]: This is causing issues because `current_instance` is not a `dict`!

```
Creator: Workfile (io.openpype.creators.blender.workfile)

Error: 'CreatedInstance' object has no attribute 'update'

Traceback (most recent call last):
  File "C:\Dev\Ynput\OpenPype\openpype\pipeline\create\context.py", line 2036, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\plugins\create\create_workfile.py", line 77, in create
    current_instance.update(
AttributeError: 'CreatedInstance' object has no attribute 'update'
```

**Comment by simonebarbieri** [comment]
> UIs will now pop up to the front if they are behind another window or were minimized - I kept searching for e.g. the loader or publisher, this resolves it in Blender.

This works! 🎉

File: openpype/hosts/blender/plugins/create/create_workfile.py (hunk 5 above)
Comment (BigRoy) [reply openpype/hosts/blender/plugins/create/create_workfile.py]: Thanks, well spotted. Fixed with https://github.com/ynput/OpenPype/pull/5910/commits/45699444ef8d4e466d3fed4652d26b4d1c55033a

Try and break it again!

File: openpype/hosts/blender/plugins/publish/extract_fbx_animation.py [hunk 6]
```diff
@@ -19,20 +22,16 @@ class ExtractAnimationFBX(publish.Extractor):
...
-        collection = [
-            obj for obj in instance if type(obj) is bpy.types.Collection][0]
-
-        # Again, the first object in the collection is taken , as there
-        # should be only the asset group in the collection.
-        asset_group = collection.objects[0]
+        asset_group = instance.data["transientData"]["instance_node"]
 
         armature = [
             obj for obj in asset_group.children if obj.type == 'ARMATURE'][0]
```
Comment (simonebarbieri) [inline openpype/hosts/blender/plugins/publish/extract_fbx_animation.py:37]: This should be something like this:

```
        collection = instance.data["transientData"]["instance_node"]
        asset_group = collection.objects[0]

        armature = [
            obj for obj in asset_group.children if obj.type == 'ARMATURE'][0]
```

Because `asset_group` is a collection, and `children` returns all the children collections.
I've tried with `asset_group.all_objects`, but it doesn't look to be working because it returns only the objects that are linked in the collections. I don't remember if this has always been the case, I think I remember that it once returned the children of linked objects even if those weren't linked, but maybe I'm not remember this correctly! In any case, it doesn't work, so it needs to be changed :)

Comment (BigRoy) [reply openpype/hosts/blender/plugins/publish/extract_fbx_animation.py]: I suspect this should do it: https://github.com/ynput/OpenPype/pull/5910/commits/d148671d02563ce11413e7cb4fd010f9fb3e41f1

It seems that:

- `Collection.objects`: Children `bpy.types.Object` of this collection, excluding child collections
- `Collection.all_objects`: All children `bpy.types.Object` including children of child collections but excluding all `bpy.types.Collection`
- `Collection.children`: Children `bpy.types.Collection` of this collection, excluding child objects
- `Collection.children_recursive`: All children`bpy.types.Collections` of this and including children of child collections recursively.

As such, none of these find the children (parenting hierarchy) of the `bpy.types.Object`. They do not traverse down the `bpy.types.Object` hierarchy. It requires a call to `Object.children` or `Object.children_recursive` to do so.

For now with this commit I opted to for all children objects of the instance collection (not looking into child collections) search the ARMATURE among all its children hierarchy of `bpy.types.Object`.

Comment (Sharkitty) [reply openpype/hosts/blender/plugins/publish/extract_fbx_animation.py]: In this case a `next` would be better. That way instead of looping through all child objects THEN choosing the first one, it stops at the first matching object found.

File: openpype/hosts/blender/plugins/publish/extract_fbx_animation.py [hunk 7]
```diff
@@ -56,7 +71,10 @@ def process(self, instance):
...
             ending_frames.append(curr_frame_range[1])
         else:
-            self.log.info("Object have no animation.")
+            self.log.info(
+                f"Armature '{armature.name}' has no animation, "
+                f"skipping FBX animation extraction for {instance}."
+            )
             return
 
         asset_group_name = asset_group.name
```
Comment (simonebarbieri) [inline openpype/hosts/blender/plugins/publish/extract_fbx_animation.py:80]: We need some more changes here! Some of them are unrelated to the PR, but it's a very quick fix that is necessary to test it so I think we can include them here. (I'll write them all here because GitHub doesn't allow me to put comments in line that don't have changes)

This is necessary because this property is here only if you load from a json layout, not from blend (it's my overlook, and I don't know how we missed before 😅)
```
        asset_name = asset_group.get(AVALON_PROPERTY).get("asset_name")
        if asset_name:
            asset_group_name = asset_group.name
            asset_group.name = asset_name
```

We have then to remove `asset_group.select_set(True)` and `asset_group.select_set(False)`, because it was an object before, but now it is a collection and it cannot be selected. We also need to remove `asset_group` in the override context (we should also change the override to use the new way to override the context, but maybe that's for another PR).
```
        # asset_group.select_set(True)
        armature.select_set(True)
        fbx_filename = f"{instance.name}_{armature.name}.fbx"
        filepath = os.path.join(stagingdir, fbx_filename)

        override = plugin.create_blender_context(
            active=armature, selected=[armature])
        bpy.ops.export_scene.fbx(
            override,
            filepath=filepath,
            use_active_collection=False,
            use_selection=True,
            bake_anim_use_nla_strips=False,
            bake_anim_use_all_actions=False,
            add_leaf_bones=False,
            armature_nodetype='ROOT',
            object_types={'EMPTY', 'ARMATURE'}
        )
        armature.name = armature_name
        if asset_name:
            asset_group.name = asset_group_name
        # asset_group.select_set(False)
        armature.select_set(False)

```

I think this is the minimum to make it work, but in the future it needs some serious rework!

**Comment by simonebarbieri** [comment]
> Animation instances in Blender seem to be much simpler than in Maya so they are easier to create; also loading rigs doesn't automatically create an animation instance yet it seems. So I believe they should remain visible.

Sorry I missed this edit! Animations are created when loading a layout with rigs, not just rigs! This was done because of the original workflow that we had in mind, that was more "Blender -> Unreal". Now it can be used to also export Alembic, which actually work within Blender, (and theoretically blend, but it doesn't work and not because of this PR), so it can be kept visible :)
In general I think the animation family needs a serious rework in the future, when we have time 😅

File: openpype/hosts/blender/plugins/publish/extract_fbx_animation.py (hunk 6 above)
Comment (BigRoy) [reply openpype/hosts/blender/plugins/publish/extract_fbx_animation.py]: Correct, but in this case I want to log if there are multiple so that's why I continue the full loop. Just so an edge case will be found easier.

File: openpype/hosts/blender/plugins/publish/extract_fbx_animation.py (hunk 7 above)
Comment (BigRoy) [reply openpype/hosts/blender/plugins/publish/extract_fbx_animation.py:80]: Does this work without selecting a "root" node? Is selecting the "armature" sufficient? I feel like we might be lacking export data then?

Comment (BigRoy) [reply openpype/hosts/blender/plugins/publish/extract_fbx_animation.py:116]: Does this work https://github.com/ynput/OpenPype/pull/5910/commits/480c37533b8ce0e542e7b052ae28ff4d7a775acb @simonebarbieri ?

Comment (simonebarbieri) [reply openpype/hosts/blender/plugins/publish/extract_fbx_animation.py:116]: It does! That's great :D Thanks!

**Comment by BigRoy** [comment]
@LiborBatek @simonebarbieri was there anything else - do we need a final test run before merging?

**Comment by LiborBatek** [comment]
> @LiborBatek @simonebarbieri was there anything else - do we need a final test run before merging?

Lets do last round of tests...will inform what are the findings asap!

File: openpype/hosts/blender/plugins/publish/validate_deadline_publish.py
[the remark this answers is not in the document]
Comment (BigRoy) [reply openpype/hosts/blender/plugins/publish/validate_deadline_publish.py:1]: Fixed with https://github.com/ynput/OpenPype/pull/5910/commits/025c114de854688bbf99e60a3d8fe3843c05025c

**Review by LiborBatek — APPROVED** [review]
I have tested both in legacy OP and Ayon mode and all good!!

Great job everyone who involved in this PR!!
